File: py/bank.py
from __future__ import annotations
from locale import currency

from money import Money


class Bank:
    def __init__(self) -> None:
        self.exchange_rates = {}

    def add_exchange_rate(self, currency_from: str, currency_to: str, rate: float):
        key = f"{currency_from}->{currency_to}"
        self.exchange_rates[key] = rate

    def convert(self, a_money: type[Money], a_currency: str) -> type[Money]:
        if a_money.currency == a_currency:
            return Money(a_money.amount, a_money.currency)

        key = f"{a_money.currency}->{a_currency}"
        if key in self.exchange_rates:
            return Money(a_money.amount * self.exchange_rates[key], a_currency)
        raise Exception(key)

File: js/bank.js
const Money = require('./money');

class Bank {
  constructor() {
    this.exchangeRates = new Map();
  }

  /**
   * Add mapping of currency
   * @param {string} currencyFrom
   * @param {string} currencyTo
   * @param {number} rate
   */
  addExchangeRate(currencyFrom, currencyTo, rate) {
    const key = currencyFrom + '->' + currencyTo;
    this.exchangeRates.set(key, rate);
  }

  /**
   * Convert money from one currency to another
   * @param {Money} money
   * @param {string} currency
   * @return {Money}
   */
  convert(money, currency) {
    if (money.currency === currency) {
      return new Money(money.amount, money.currency);
    }
    const key = money.currency + '->' + currency;
    const rate = this.exchangeRates.get(key);
    if (rate === undefined) {
      throw new Error(key);
    }
    return new Money(money.amount * rate, currency);
  }
}

module.exports = Bank;
